---
title: "exprAnalysis-Vignette"
author: "Shirin Glander"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: spacelab
    highlight: kate
  pdf_document:
    toc: yes
vignette: >
  %\VignetteIndexEntry{exprAnalysis-Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

-----------

<br>

# exprAnalysis package

This package combines functions from various packages used to analyze and visualize expression data from NGS or expression chips.

  + It supports normalized input as e.g. from Cufflinks or expression chip arrays and raw count data from bam file input.
  + It supports mRNA, miRNA, protein or other expression data.

So far, it is implemented for human and mouse data!

It uses function from the following packages:

  + AnnotationDbi for annotating gene information
  + beadarray for importing Illumina expression chip files from GenomeStudio
  + clusterProfiler and DOSE for functional enrichment analysis
  + DESeq2 for differential expression analysis of raw count data
  + GenomicAlignments, GenomicFeatures, Rsamtools for reading bam files
  + pcaGoPromoter for principle component analysis
  + limma for differential expression analysis of normalised expression data
  + pathview for mapping KEGG pathways
  + pheatmap for heatmaps
  + sva for batch correction
  + WGCNA for coregulatory network determination

-----------

<br>

# Prepare session

```{r library_private, echo=FALSE, include=FALSE, cache=FALSE}
#require(devtools)
devtools::load_all()
```

```{r library_official, eval=FALSE}
# update all packages
source.bioc="https://bioconductor.org/biocLite.R"
source(source.bioc)
biocLite("BiocUpgrade")
biocLite(ask=FALSE) 

update.packages()

# make sure the workspace is in pristine condition
rm(list=ls(all=TRUE))

orig_par <- par(no.readonly=T)

options(stringsAsFactors = FALSE)
enableWGCNAThreads()
```

-----------

<br>

## The package can be installed via github:

Beware that the vignette is rather large and thus takes a minute to compile. You can also see the Vignette at [https://github.com/ShirinG/exprAnalysis/blob/master/vignettes/exprAnalysis.Rmd](https://github.com/ShirinG/exprAnalysis/blob/master/vignettes/exprAnalysis.Rmd).

```{r library_install, eval=FALSE}
# install package from github
# either the latest stable release that passed TRAVIS CI check
devtools::install_github("ShirinG/exprAnalysis", build_vignettes=TRUE, ref = "stable.version0.1.0")

# or the development version
devtools::install_github("ShirinG/exprAnalysis", build_vignettes=TRUE, ref = "master")

library("exprAnalysis")

# To view the vignette if you built it with the package:

vignette("exprAnalysis", package="exprAnalysis")
vignette("CummeRbund", package="exprAnalysis")

browseVignettes("exprAnalysis")

# There might be problems with installation of packages. In order to install them manually:
list.of.packages <- c("list", "of", "packages")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]

# CRAN
if(length(new.packages)>0) install.packages(new.packages)

# Bioconductor
if(length(new.packages)>0) {
  source("https://bioconductor.org/biocLite.R")
  biocLite(new.packages)
}

```

-----------

<br>

# Functions

See package help (?functionname, ?data) for detailed description of functions and example datasets.

-----------

<br>

# Input data

Takes an expression data matrix containing numeric values as expression measures (e.g. read count data, FPKM values, Illumina expression data).

- Rownames should be gene or isoform identifiers (e.g. gene names)
- Colnames should be sample IDs (sample names)

This package contains two example matrices of randomly generated expression data as raw read counts (called **"countmatrix"**) and as normalized read counts (called **"expmatrix"**).

```{r input, include=FALSE}
#expmatrix <- read.table(system.file("extdata", "Random_exprMatrix.txt", package = "exprAnalysis"), header = TRUE, sep = "\t")
# doesn't exist any more (is now stored as raw data)
```

```{r input_data, eval=FALSE}
data("countmatrix")
data("expmatrix")
```

-----------

<br>

## Starting from FPKM data

If you have FPKM data (e.g. quantile normalized) from Cufflinks, treat the data such as the **expmatrix** example.

<br>

### Cuffdiff

See CummeRbund vignette CummeRbundVignette.html.

-----------

<br>

## Starting from count data

If you want to do count data analysis, you can either produce a count matrix (e.g. with HTSeq) and proceed to DESeq2 analysis or you can produce the count table directly from the bam files and then proceed to DESeq2. 

read_bam_to_countmatrix() returns a DESeq data set that can directly be used with the DEseq2 pipeline. Or you can manually load the count matrix that was saved by read_bam_to_countmatrix() and then go to DESeqDataFrameFromMatrix().

In read_bam_to_countmatrix() fragments will be counted only once to each gene, even if they overlap multiple exons of a gene which may themselves be overlapping.

```{r countdata_bam, eval=FALSE}
# Locate alignment files
dir <- getwd()
filenames <- fileslist[grep(".*sorted.bam$", list.files(dir))]

# Create a sample table

samplename <- sub("_accepted_hits.sorted.bam", "", filenames)
design <- c("Treatment", "Control")
sampleid <- sub("Sample", "", samplename)

sampleTable <- data.frame(row.names = samplename, sampleid = sampleid, filenames = filenames, colData = design, stringsAsFactors = F)

#         sampleid                         filenames    colData
#Sample1         1  Sample1_accepted_hits.sorted.bam  Treatment
#Sample2         2  Sample2_accepted_hits.sorted.bam    Control

data <- read_bam_to_countmatrix(sampleTable, gtffile = "Homo_sapiens.GRCh38.83.gtf", projectfolder = getwd(), outPrefix="Test")

countmatrix <- SummarizedExperiment::assay(data)

```

<br>

### Continue count data analysis with DESeq2

See DESeq2 Vignette for additional details.

#### If you do not directly proceed from read_bam_to_countmatrix():

- read in saved count matrix
- define experimental design
- convert to DESeq data set

```{r read_in_countdata, eval=FALSE}
countmatrix <-   read.table(file.path(projectfolder, "Countdata", outPrefix, paste0(outPrefix, "_Summarize_overlaps_count_matrix.txt")), header=T, sep = "\t")
```

```{r read_in_countdata2, eval=FALSE}
design <- gsub("(.*)(_[0-9])", "\\1", colnames(countmatrix))
ExpDesign <- data.frame(row.names=colnames(countmatrix), treatment = design)

data <- DESeq2::DESeqDataSetFromMatrix(countData = countmatrix, colData=ExpDesign, design=~treatment)
```

<br>

### DESeq2

- optional, but recommended: remove genes with zero counts over all samples
- run DESeq
- Extracting transformed values

Note: the rlog transformation is provided for applications other than differential testing. For differential testing we recommend the DESeq function applied to raw counts, as described later in this workflow, which also takes into account the dependence of the variance of counts on the mean value during the dispersion estimation step.


```{r DESeq2, eval=FALSE}
data <- data[rowSums(DESeq2::counts(data)) > 1, ]

data_DESeq <- DESeq2::DESeq(data)

expmatrix_DESeq <- DESeq2::rlog(data_DESeq, fitType="local")
expmatrix <- SummarizedExperiment::assay(expmatrix_DESeq)
```

<br>

#### Dispersion plot

```{r dispersion_plot, eval=FALSE, fig.width=10, fig.height=5}
DESeq2::plotDispEsts(data_DESeq, main="Dispersion Estimates")
```

-----------

<br>

## Starting from Affymetrix expression chips

If you have CEL files, start with the following code to produce the expression matrix and then treat it like the **expmatrix** example:

- Read in the data and create an expression, using RMA for example

Currently the rma function implements RMA in the following manner
1. Probe specific correction of the PM probes using a model based on observed intensity being the sum of signal and noise
2. Normalization of corrected PM probes using quantile normalization (Bolstad et al., 2003)
3. Calculation of Expression measure using median polish.

```{r Affymetrix, eval=FALSE}
library(affy)

# Choose directory containing CEL files
dir <- getwd()

# check, that you have the correct directory
celfiles <- list.files(dir, recursive = TRUE)[grep(".*CEL$", list.files(dir, recursive = TRUE), ignore.case = TRUE)]

affydata <- ReadAffy(filenames=celfiles)
MAplot(affydata,plot.method="smoothScatter")

eset <- rma(affydata)

quality_control_plots(eset, groupColumn=character(0))

# Optional: Remove bad quality probes/ genes and/ or samples

write.exprs(eset, file="eset.txt")

library(made4)
overview(eset)

# RNA degradation plots
deg <- AffyRNAdeg(affydata)
plotAffyRNAdeg(deg)


expmatrix <- exprs(eset)
```

-----------

<br>

## Starting from Illumina expression chips

If you have Illumina intensity data, load it into GenomeStudio first:

- Open New Project -> Gene Expression
- Choose Assay Type -> Next
- Choose Project Repo and name the project -> Next
- On the left side, mark all folders containing data, select "All" and import them into the right window -> Next
- Name Group Set, on the left side, mark all folders containing data, select "All" and import them into the right window (optionally, you can directly create groups here) -> Next
- For now, don't use normalisation and don' substract background
- Choose appropriate Content Descriptor (e.g. HumanHT-12_V4_0_R1_15002873_B.bgx) -> Finish
- Export SampleProbeProfile.txt: Choose columns AVG_Signal, Detection Pval, BEAD_STDERR and Avg_NBEADS
- Export ControlProbeProfile.txt: Choose columns AVG_Signal and Detection Pval
- You also need the Samplesheet.csv

```{r Illumina, eval=FALSE, include=FALSE}
dataFile      <- file.path("U:", "exprAnalysis_test/SampleProbeProfile.txt")
qcFile        <- file.path("U:", "exprAnalysis_test/ControlProbeProfile.txt")
sampleSheet   <- file.path("U:", "exprAnalysis_test/samplesheet.csv")
```

```{r Illumina_official, eval=FALSE}
dataFile      <- file.path("SampleProbeProfile.txt")
qcFile        <- file.path("ControlProbeProfile.txt")
sampleSheet   <- file.path("samplesheet.csv")

# define Illumina Expression Array: "HumanHT-12 v?", "MouseWG-6 v?", "MouseRef-8 v?"
expressionchipType <- "HumanHT-12 v4"

eset <- read_Illumina(dataFile, qcFile, sampleSheet, expressionchipType, ProbeID = "PROBE_ID", skip = 0, controlID="ProbeID", qc.skip = 0, method_norm ="none", transform="log2")

quality_control_plots(eset)

# Optional: Remove bad quality probes/ genes and/ or samples

eset <- normalise_eset(eset)

expmatrix <- exprs(eset)
```

-----------

<br>

# Batch correction

```{r batch_cor, eval=FALSE}
pheno <- data.frame(sample=c(1:16), treatment=sub("(.*)(_[0-9])", "\\1", colnames(expmatrix)), batch=ifelse(grepl("Ctrl", colnames(expmatrix)) == TRUE, "1", ifelse(grepl("ActLPS", colnames(expmatrix)) == TRUE, "1", "2")), row.names = colnames(expmatrix))

expmatrix <- batch_removal(expmatrix, pheno)
```

-----------

<br>

# Exploratory analysis of all genes

## Variance vs mean gene expression across samples

Plots variance against mean gene expression across samples and calculates the correlation of a linear regression model.

```{r var_vs_mean, fig.width=6, fig.height=5}
var_vs_mean(countmatrix)
# var_vs_mean(expmatrix)
```

-----------

<br>

## Intersample variances

```{r variance_sumOverlaps, eval=FALSE, fig.width= 5}
library(corrgram)

Ctrl_cor <- expmatrix[,grep("Ctrl", colnames(expmatrix))]

corrgram::corrgram(Ctrl_cor, order=TRUE, lower.panel=corrgram::panel.pie,
         upper.panel=corrgram::panel.pts, text.panel=corrgram::panel.txt,
         main="Correlogram of controls")
```

-----------

<br>

## Principle Component Analysis

Uses functions as implemented by **"pcaGoPromoter"**.

You can only plot the principle components using:

```{r pca_plot, eval=TRUE, fig.width=6, fig.height=6}
groups <- as.factor(c(rep("Ctrl",4), rep("TolLPS",4), rep("TolS100A8",4), rep("ActLPS",4)))

pca_plot(expmatrix, groups)
```

<br>

Or you can plot the principle components and calculate TF and GO term enrichments of genes (defaults to top 2.5%) with highest and lowest loadings. With this function, the ouput files are directly saved to .pdf and .txt (by default to working directory).

```{r pca_plot_enrich, eval=FALSE}
pca_plot_enrich(expmatrix, groups)
```

-----------

<br>

## Heatmaps

Here, of the 30 most highly expressed genes.

```{r heatmaps, eval=TRUE}

select <- order(rowMeans(expmatrix),decreasing=TRUE)[1:30]

heatmaps(expmatrix[select,], samplecols = rep(c("#E41A1C", "#377EB8", "#4DAF4A", "#984EA3"), each=4))
```

```{r heatmaps2, eval=FALSE}
# Heatmap function from DESeq2, using pheatmap:

library(pheatmap)

sampleDists <- dist(t(expmatrix))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(expmatrix_DESeq$treatment)
colnames(sampleDistMatrix) <- NULL
colors <- grDevices::colorRampPalette( rev(RColorBrewer::brewer.pal(9, "Blues")) )(255)
pheatmap::pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

df <- data.frame(treatment = SummarizedExperiment::colData(data_DESeq)[,c("treatment")], row.names = rownames(SummarizedExperiment::colData(data_DESeq)))
pheatmap::pheatmap(expmatrix[select,], cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df)
```

```{r int_heatmaps, eval=FALSE}
# Interactive heatmap

#devtools::install_github('talgalili/heatmaply')
library(heatmaply)

# sample correlation
hm <- heatmapr(cor(expmatrix[select,]), k_row = 4, k_col = 4)

# gene correlation
hm <- heatmapr(cor(t(expmatrix[select,])), k_row = NULL, k_col = NULL)

heatmaply(hm)

#Hover over heatmap to see individual values and sample/ gene IDs
```

-----------

<br>

## WGCNA

<br>

### Hierarchical Clustering and outlier detection

Uses adjacency matrix function from **"WGCNA"** and hierarchical clustering from **"flashClust"**.

```{r wgcna_sample_dendrogram, eval=TRUE, fig.width=8, fig.height=5}
datTraits <- data.frame(Ctrl = c(rep(1, 4), rep(0,12)), TolPS = c(rep(0, 4), rep(1, 4),rep(0, 8)), TolS100A8 = c(rep(0, 8), rep(1, 4), rep(0, 4)), ActLPS = c(rep(0, 12),rep(1, 4)), Tol = c(rep(0, 4), rep(1, 8), rep(0, 4)), ExPhenotype = c(stats::rnorm(4, 10, 1),stats::rnorm(8, 25, 1),stats::rnorm(4, 50, 1)), row.names = colnames(expmatrix))

datExpr <- wgcna_sample_dendrogram(expmatrix, datTraits)

# Optional: Remove outlier samples and repeat
```

<br>

### Choosing a Soft Threshold

Ideally, pick a SFT with R^2 > 0.9. In this example, this threshold was not reached, so I pick the highest SFT: 30.

```{r wgcna_sft, eval=FALSE}
# Choose a set of soft thresholding powers
powers=c(1:30)

# choose power based on SFT criterion
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5, networkType ="signed", corFnc= "bicor", corOptions=list(maxPOutliers=0.1))

# Plot the results:
tiff(filename= "SFT.tiff", width = 8000 , height = 4000, res=600, compression = "lzw")
par(mfrow=c(1,2))
# SFT index as a function of different powers
plot(sft$fitIndices[,1],-sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     xlab="Soft Threshold (power)",ylab="SFT, signed R^2",type="n",main=paste("Scale independence"))
text(sft$fitIndices[,1],-sign(sft$fitIndices[,3])*sft$fitIndices[,2],
     labels=powers,col="red")
# this line corresponds to using an R^2 cut-off of h
abline(h=0.90,col="red")
# Mean connectivity as a function of different powers
plot(sft$fitIndices[,1],sft$fitIndices[,5],type="n",
     xlab="Soft Threshold (power)",ylab="Mean Connectivity",main=paste("Mean connectivity"))
text(sft$fitIndices[,1],sft$fitIndices[,5],labels=powers,col="red")
dev.off()
```

<br>

### Automatic module detection via dynamic tree cutting

```{r wgcna_net, eval=FALSE}
softpower = 30
mergingThresh = 0.25
net = blockwiseModules(datExpr,checkMissingData = TRUE,corType="bicor", # or pearson
                       maxBlockSize=10000,networkType="signed",power=softpower,
                       minModuleSize=30,maxPOutliers = 0.1,mergeCutHeight=mergingThresh,
                       numericLabels=TRUE,saveTOMs=TRUE,randomSeed = 12345,
                       pamRespectsDendro=FALSE,saveTOMFileBase="TESTexprsTOM")
str(net)
```

```{r wgcna_moduleColors, eval=FALSE}
moduleColors <- wgcna_plotDendroAndColors(datExpr, datTraits, net)
```

<br>

```{r wgcna_MEs, eval=FALSE}
# Recalculate MEs with color labels
MEs0 = moduleEigengenes(datExpr, moduleColors, softPower = softpower)$eigengenes
# Reorder given (eigen-)vectors such that similar ones (as measured by correlation) are next to each other:
MEs = orderMEs(MEs0)
rownames(MEs) <- rownames(datExpr)
write.table(MEs, "Automatic_ModuleEigengenes_numberLabel.txt", row.names=T, quote=F, sep="\t")

datKME <- wgcna_modulememberships(datExpr, MEs, moduleColors)

wgcna_heatmaps(MEs, datExpr, datTraits)
```

![Module trait correlation heatmap](wgcnaHeatmap.png)

<br>

### Output file for gene ontology analysis

```{r wgcna_output, eval=FALSE}
# Annotate
genes = colnames(datExpr)
datExpr_anno <- geneAnnotations(input=genes, keys=genes, column=c("ENTREZID", "ENSEMBL"), keytype="SYMBOL", organism = "human")

# Correlations of genes and traits
traitsCor=cor(datExpr,datTraits,use="p")
colnames(traitsCor)=paste("cor",colnames(traitsCor),sep=".")

dataOutput=data.frame(datExpr_anno ,moduleColors, datKME, traitsCor)

write.table(dataOutput,"AllGenesResults.txt",row.names=F,sep="\t")
```

<br>

### Module eigengene expression plots

```{r wgcna_data, eval=FALSE}
data(MEs)
```

```{r module_eigengene_plot, eval=TRUE, fig.width=10, fig.height=15}
layout(matrix(c(1:length(colnames(MEs)),12,rep(13,3)), ncol=3, byrow=TRUE), heights=c(rep(3,4), 0.5))

for (color in gsub("ME", "", colnames(MEs))){
  module_eigengene_plot(groups, MEs, color)
}
plot.new()

par(mai=c(0,0,0,0))
plot.new()
legend("bottom", c("mean module eigengenes", "replicate module eigengene values"), xpd = TRUE, horiz = TRUE, inset = c(0,0), bty = "n", col = c("grey", "black"), pch = 16, cex = 2)

```

-----------

<br>

# Differential expression analysis

## DE analyis using limma

For normalized expression data (FPKM from Cufflinks or intensities from expression chip arrays).

```{r diff_limma, eval=FALSE, warning=FALSE}
design <- as.matrix(data.frame(Ctrl = c(rep(1, 4), rep(0,12)), TolLPS = c(rep(0, 4), rep(1, 4),
              rep(0, 8)), TollMRP8 = c(rep(0, 8), rep(1, 4), rep(0, 4)), ActLPS = c(rep(0, 12),
              rep(1, 4)), row.names = colnames(expmatrix)))
groupcomparisons=c("TolLPS-Ctrl", "TollMRP8-Ctrl", "ActLPS-Ctrl")
DEgenes_all <- diff_limma_all(expmatrix, design, groupcomparisons)

comparison="TolLPS-Ctrl"
Allgenes_limma_pw <- diff_limma_pw_unfiltered(expmatrix, design, comparison)

DEgenes_pw <- Allgenes_limma_pw[which(abs(Allgenes_limma_pw$logFC) > log2(1.5) & Allgenes_limma_pw$adj.P.Val < 0.05),]
```

Example output files from diff_limma functions have been saved to the package data and can be accessed via:

```{r diff_limma_output, eval=FALSE}
data("DEgenes_pw")
data("Allgenes_limma_pw")


# Export tables to Latex
library(xtable)

italic <- function(x){
paste0('{\\emph{ ', x, '}}')
}

print(xtable(DEgenes_pw[1:10,], 
       digits = 2,
       caption = c("\\tt Long caption", "Short caption"),
       label = "tab:testtable",
       auto = TRUE),
      caption.placement = "top",
      sanitize.rownames.function = italic,
      booktabs = TRUE,
      floating = TRUE, 
      latex.environments = "center",
      include.rownames = TRUE,
      scalebox = 0.8,
      tabular.environment = "tabularx", width = "\\textwidth")

# or to html
print(xtable(DEgenes_pw[1:10,], 
       digits = 2,
       caption = c("Long caption", "Short caption"),
       label = "tab:testtable",
       auto = TRUE), type = "html", caption.placement = "top")
```

<!-- html table generated in R 3.3.0 by xtable 1.8-2 package -->
<!-- Wed Jun 01 08:44:34 2016 -->
<table border=1>
<caption align="top"> Long caption </caption>
<tr> <th>  </th> <th> logFC </th> <th> CI.L </th> <th> CI.R </th> <th> AveExpr </th> <th> t </th> <th> P.Value </th> <th> adj.P.Val </th> <th> B </th> <th> ENTREZID </th> <th> ENSEMBL </th>  </tr>
  <tr> <td> HCAR2 </td> <td align="right"> -1.60 </td> <td align="right"> -1.78 </td> <td align="right"> -1.41 </td> <td align="right"> 5.17 </td> <td align="right"> -18.38 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 16.91 </td> <td> 338442 </td> <td> ENSG00000182782 </td> </tr>
  <tr> <td> LOC100996342 </td> <td align="right"> 2.54 </td> <td align="right"> 2.18 </td> <td align="right"> 2.90 </td> <td align="right"> 6.89 </td> <td align="right"> 15.03 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 14.33 </td> <td> 100996342 </td> <td> ENSG00000235478 </td> </tr>
  <tr> <td> INPP4B </td> <td align="right"> -1.54 </td> <td align="right"> -1.77 </td> <td align="right"> -1.31 </td> <td align="right"> 4.80 </td> <td align="right"> -14.19 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 13.58 </td> <td> 8821 </td> <td> ENSG00000109452 </td> </tr>
  <tr> <td> LOC100505549 </td> <td align="right"> -1.26 </td> <td align="right"> -1.45 </td> <td align="right"> -1.06 </td> <td align="right"> 3.11 </td> <td align="right"> -13.91 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 13.32 </td> <td> 100505549 </td> <td>  </td> </tr>
  <tr> <td> FCGBP </td> <td align="right"> -2.06 </td> <td align="right"> -2.38 </td> <td align="right"> -1.75 </td> <td align="right"> 3.39 </td> <td align="right"> -13.87 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 13.28 </td> <td> 8857 </td> <td> ENSG00000275395 </td> </tr>
  <tr> <td> ZSCAN31 </td> <td align="right"> -1.40 </td> <td align="right"> -1.63 </td> <td align="right"> -1.18 </td> <td align="right"> 3.44 </td> <td align="right"> -13.33 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 12.76 </td> <td> 64288 </td> <td> ENSG00000235109 </td> </tr>
  <tr> <td> MIR3665 </td> <td align="right"> -1.32 </td> <td align="right"> -1.53 </td> <td align="right"> -1.11 </td> <td align="right"> 2.69 </td> <td align="right"> -13.27 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 12.70 </td> <td> 100500861 </td> <td> ENSG00000266325 </td> </tr>
  <tr> <td> PTH2R </td> <td align="right"> -1.62 </td> <td align="right"> -1.90 </td> <td align="right"> -1.35 </td> <td align="right"> 3.69 </td> <td align="right"> -12.56 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 11.98 </td> <td> 5746 </td> <td> ENSG00000144407 </td> </tr>
  <tr> <td> GCLC </td> <td align="right"> -2.42 </td> <td align="right"> -2.83 </td> <td align="right"> -2.00 </td> <td align="right"> 3.13 </td> <td align="right"> -12.27 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 11.66 </td> <td> 2729 </td> <td> ENSG00000001084 </td> </tr>
  <tr> <td> SOX30 </td> <td align="right"> 2.54 </td> <td align="right"> 2.10 </td> <td align="right"> 2.98 </td> <td align="right"> 8.57 </td> <td align="right"> 12.19 </td> <td align="right"> 0.00 </td> <td align="right"> 0.00 </td> <td align="right"> 11.59 </td> <td> 11063 </td> <td> ENSG00000039600 </td> </tr>
   </table>
  
-----------

<br>

## DE analyis using DESeq2

For raw read count data.

##### contrast DE groups:
 + lfc = treatment > Ctrl, - lfc = treatment < Ctrl
 p-value & p.adjust values of NA indicate outliers detected by Cook's distance
 NA only for p.adjust means the gene is filtered by automatic independent filtering for having a low mean normalized count

<br>

```{r DEgenes_DESeq2, eval=FALSE, warning=FALSE}
library(DESeq2)
library(ggplot2)
library(ggrepel)

# find possible contrasts with
DESeq2::resultsNames(data_DESeq)

res <- DESeq2::results(data_DESeq, contrast=list("treatmentActLPS", "treatmentCtrl"), cooksCutoff = 0.99, independentFiltering = TRUE, alpha = 0.05, pAdjustMethod = "BH")

# order results table by the smallest adjusted p value:
res <- res[order(res$padj),]

results = as.data.frame(dplyr::mutate(as.data.frame(res), sig=ifelse(res$padj<0.05, "FDR<0.05", "Not Sig")), row.names=rownames(res))

DEgenes_DESeq <- results[which(abs(results$log2FoldChange) > log2(1.5) & results$padj < 0.05),]

p = ggplot2::ggplot(results, ggplot2::aes(log2FoldChange, -log10(pvalue))) +
  ggplot2::geom_point(ggplot2::aes(col=sig)) +
  ggplot2::scale_color_manual(values=c("red", "black")) +
  ggplot2::ggtitle("Volcano Plot of DESeq2 analysis")

p+ggrepel::geom_text_repel(data=results[1:10, ], ggplot2::aes(label=rownames(results[1:10, ])))

# If there aren't too many DE genes:
#p+geom_text_repel(data=dplyr::filter(results, padj<0.05), aes(label=rownames(results[1:10, ])))

```

```{r volcano_simple, eval=FALSE, include=FALSE, warning=FALSE}
# Make a basic volcano plot
plot(res[,2], -log10(res[,6]), main = "Volcano plot", xlab="l2fc", ylab="-log10(q-val)", pch = 20, col=scales::alpha("black", 0.5))

# Add colored points: green if padj<0.05, orange of log2FC>1, red if both)
points(res[which(res$padj<.0001),2], -log10(res[which(res$padj<.0001),6]), pch=16, col=scales::alpha("green", 0.5))
points(res[which(abs(res$log2FoldChange)>2),2], -log10(res[which(abs(res$log2FoldChange)>2),6]), pch=16, col=scales::alpha("orange", 0.5))
points(res[which(abs(res$log2FoldChange)>2 & res$padj<.0001),2], -log10(res[which(abs(res$log2FoldChange)>2 & res$padj<.0001),6]), pch=16, col=scales::alpha("red", 0.5))

abline(h=-log10(.0001), lty=2, lwd=2)
abline(v=2, lty=3, lwd=2)
abline(v=-2, lty=3, lwd=2)

calibrate::textxy(res[1:10,2], -log10(res[1:10,6]), labs=rownames(res[1:10,]), pos=3)
```

<br>

#### MA-plot

```{r ma_plot, eval=FALSE, fig.width=5, fig.height=5}
DESeq2::plotMA(res, main="MA Plot", ylim=c(-2,2))
```

<br>

#### plotCounts, which normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting

```{r plot_counts, eval=FALSE, fig.width=10, fig.height=3}
par(mfrow=c(1,3))

for (i in 1:3){
  gene <- rownames(res)[i]
  main = gene
  DESeq2::plotCounts(data_DESeq, gene=gene, intgroup="treatment", main = main)
}
```

-----------

<br>

## Gene annotations

Can be used to add e.g. ENTREZ ID, ENSEMBL ID, etc. to gene name.

```{r geneAnnotations, eval=TRUE, message=FALSE, warning=FALSE}
DEgenes_pw <- geneAnnotations(input=DEgenes_pw, keys=row.names(DEgenes_pw), column=c("ENTREZID", "ENSEMBL"), keytype="SYMBOL", organism = "human")
```

```{r geneAnnotations2, eval=FALSE, message=FALSE, warning=FALSE}
DEgenes_DESeq <- geneAnnotations(input=DEgenes_DESeq, keys=row.names(DEgenes_DESeq), column=c("ENTREZID", "ENSEMBL"), keytype="SYMBOL")
```
-----------

<br>

## Enrichment Analysis using clusterPofiler

```{r clusterProfiler, eval=FALSE, warning=FALSE, message=FALSE}
library(clusterProfiler)
library(DOSE)

geneList <- as.vector(Allgenes_limma_pw$logFC)
names(geneList) <- rownames(Allgenes_limma_pw)
gene <- na.omit(DEgenes_pw$ENTREZID)

OrgDb <- org.Hs.eg.db::org.Hs.eg.db # can also be org.Mm.eg.db::org.Mm.eg.db

# Group GO
ggo <- clusterProfiler::groupGO(gene     = gene,
                                OrgDb    = OrgDb,
                                ont      = "BP",
                                level    = 3,
                                readable = TRUE)
head(summary(ggo))
barplot(ggo, drop=TRUE, showCategory=12)

# GO over-representation test
ego <- clusterProfiler::enrichGO(gene          = gene,
                                 OrgDb         = OrgDb,
                                 ont           = "BP",
                                 pAdjustMethod = "BH",
                                 pvalueCutoff  = 0.05,
                                 qvalueCutoff  = 0.05, 
                                 readable      = TRUE)
head(summary(ego))
barplot(ego, showCategory=25)
clusterProfiler::dotplot(ego, showCategory=25)
#enrichMap(ego)
cnetplot(ego, categorySize="pvalue", foldChange=geneList)
clusterProfiler::plotGOgraph(ego)

# enrichGO test the whole GO corpus and enriched result may contains very general terms. With dropGO function, user can remove specific GO terms or GO level from results obtained from both enrichGO and compareCluster.


## KEGG over-representation test
kk <- clusterProfiler::enrichKEGG(gene         = gene,
                 organism     = 'hsa',
                 pvalueCutoff = 0.05)
head(summary(kk))
barplot(kk, showCategory=8)
clusterProfiler::dotplot(kk)
cnetplot(kk, categorySize="geneNum", foldChange=geneList)

```


-----------

<br>

## Pathview

For Toll-like-receptor signaling:

```{r pathview, eval=FALSE}
pathview_func(DEgenes_pw, logFCcolumn="logFC", pathway.id = "04620", out.suffix = "DE_TolLPS")
```

-----------

<br>

## Venn diagrams and Biological theme comparison

```{r venn, eval=FALSE, fig.width=5, fig.height=5}
library(gplots)

venn_list <- list(genelist1 = na.omit(rownames(DEgenes_pw)[1:50]), genelist2 = na.omit(rownames(DEgenes_pw)[30:80]), genelist3 = na.omit(rownames(DEgenes_pw)[48:100]))

gplots::venn(venn_list, show.plot=TRUE, small=0.7, showSetLogicLabel=FALSE, simplify = TRUE)

mtext("If you want a header, put it here", side=3, cex = 0.8)
```

```{r compareCluster, eval=FALSE}
# Biological theme comparison
compGO <- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                          fun           = "enrichGO",
                                          ont           = "BP",
                                          OrgDb         = OrgDb,
                                          qvalueCutoff  = 0.001,
                                          pAdjustMethod = "BH",
                                          readable      = TRUE)


compKEGG <- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                            fun           = "enrichKEGG",
                                            qvalueCutoff  = 0.05,
                                            pAdjustMethod = "BH")
compKEGG <- clusterProfiler::setReadable(compKEGG, OrgDb = OrgDb)


compMKEGG <- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                             fun           = "enrichMKEGG",
                                             organism='hsa', minGSSize=1)

compDO <- clusterProfiler::compareCluster(geneCluster   = venn_list,
                                          fun           = "enrichDO",
                                          qvalueCutoff  = 0.05,
                                          pAdjustMethod = "BH",
                                          readable      = TRUE)

# For visualisation see Enrichment Analysis using clusterProfiler
```

-----------

<br>

# Networks

## TF networks

Organism can be human or mouse. For human, the longer published list of TFs is used; for mouse the shorter list provided by Bonn (for which I don't have any more info on where it comes from).

This function produced two output files:

- The .expression matrix of TF expression data. 

It will have to be opened with Biolayout3D: 
Set Minimum Correlation and Correlation metric (by default 0.7 and Pearson). Then choose a suitable correlation coefficient (Graph Degree Distribution should be close to linear). Save the resulting network as a TGF file.

Open the TGF file in Cytoscape: Open network from file. Then got to Advanced Options: untick "Use first column names" and add " to "Other:". Now set Column 2 as Source and Column 5 as Target.

- Open the node annotation table in .txt format.

In cytoscape, open table from file, import data as Node Table Columns. You can then customize the look of your network.

For example, go to Style and 

- go to Border Paint, Column: DE, Mapping type: Discrete and choose a color for 1 and 0

- go to Fill Color, Column logFC, Mapping Type: Continuous and choose colors.

```{r networks, eval=FALSE}
TF_networks(expmatrix, nodeAnno=Allgenes_limma_pw)
```

<div style="width:800px; height=800px">
![TF network Pearson 0.7](CopyOfnetwork_test.png)
</div>

-----------

<br>

## BiNGO (Cytoscape)

- Go to Cytoscape

- Apps -> BiNGO

- The BiNGO Settings panel pops up. Start by filling in a name for your cluster. This name will be used for creating the output file and the visualization of the results in Cytoscape.

- Check Paste Genes from Text and paste gene names into field below

- We want to assess overrepresentation of GO categories, and we want to visualize the results in Cytoscape. The corresponding boxes are checked accordingly by default. Then select a statistical test (the Hypergeometric Test is exact and equivalent to an exact Fisher test, the Binomial Test is less accurate but quicker) and a multiple testing correction (we recommend Benjamini & Hochberg's FDR correction, the Bonferroni correction will be too conservative in most cases), and choose a significance level, e.g. 0.05. Since we only want to visualize those GO categories that are overrepresented after multiple testing correction, and their parents in the GO hierarchy, select the corresponding visualization option. We're interested in assessing the overrepresentation of functional categories in our cluster with respect to the whole genome, which is why we choose the Complete Annotation as the reference set. 

- Select GO _Biological _Process from the ontology list, and Human from the organism list. We want to consider all evidence codes, so don't fill in anything in the evidence code box. 

- Finally, select a directory to save the output file in (the file will be named test.bgo if you filled in test as a cluster name), and press Start BiNGO...
 
- The program will inform you of its progress while parsing the annotations and calculating the tests, corrections and layout. Finally, a visualization of the overrepresented GO categories is created in Cytoscape. Uncolored nodes are not overrepresented, but they are the parents of overrepresented categories further down. Yellow nodes represent GO categories that are overrepresented at the significance level . For more significant p-values, the node color gets increasingly more orange (see also the Color Legend panel). If you'd like another layout, e.g. hierarchical, select the corresponding option from the Cytoscape Visualization menu. Regardless of the layout you choose, you'll probably have to tweak the nodes a little in order to avoid overlapping node labels. The list of significantly overrepresented functional categories is shown in the BiNGO output window (more information on the cluster and options you selected, and on which genes did not produce any annotation, is stored in the test.bgo file).

-----------

<br>

# Other functions

## miRNA target identification

```{r multiMiR, eval=FALSE}
# The current version of multiMiR is 1.0.1 and can be downloaded from
http://multimir.ucdenver.edu/multiMiR_1.0.1.tar.gz.

install.packages("XML")
install.packages("RCurl")
install.packages("/pathname/multiMiR_1.0.1.tar.gz", repos=NULL, type="source")

library(multiMiR)

miRNA_list <- as.character(c("hsa-miR-146b-3p", "hsa-miR-155-5p", "hsa-miR-4521"))

miRNA_allTargets = get.multimir(org="hsa", mirna=miRNA_list, target=NULL, table="all", summary=TRUE, predicted.cutoff.type="p", predicted.cutoff=NULL)
str(miRNA_allTargets)
    
write.table(miRNA_allTargets$predicted, "miRNAs_allTargets_predictedTargets.txt", row.names = F, col.names = T, sep="\t")
write.table(miRNA_allTargets$validated, "miRNAs_allTargets_validatedTargets.txt", row.names = F, col.names = T, sep="\t")
write.table(miRNA_allTargets$disease.drug, "miRNAs_allTargets_diseaseDrugTargets.txt", row.names = F, col.names = T, sep="\t")
write.table(miRNA_allTargets$summary, "miRNAs_allTargets_summaryTargets.txt", row.names = F, col.names = T, sep="\t")
```

-----------

<br>

## Hypergeometric test for enrichment

Analogous to Fisher's Exact Test.

```{r phyper, eval=FALSE}
phyper(SampleSuccesses -1, PopulationSuccesses, PopulationSize - PopulationSuccesses, sampleSize, lower.tail = FALSE)
```

-----------

<br>

```{r render, eval=FALSE, echo=FALSE}
rmarkdown::render('vignettes/exprAnalysis.Rmd', output_file = 'U:/exprAnalysis_test/exprAnalysisVignette.html')
```


## Session Info

```{r sessionInfo, echo=TRUE}
sessionInfo()
```